stand/efi/loader/framebuffer.c

```diff
@@ -650,17 +651,22 @@ efi_find_framebuffer(teken_gfx_t *gfx_state)
 	gfx_state->tg_fb.fb_bpp = fls(efifb.fb_mask_red | efifb.fb_mask_green |
 	    efifb.fb_mask_blue | efifb.fb_mask_reserved);
 
-	if (gfx_state->tg_shadow_fb != NULL)
-		BS->FreePages((EFI_PHYSICAL_ADDRESS)gfx_state->tg_shadow_fb,
+	/*
+	 * EFI_PHYSICAL_ADDRESS is 8 bytes wide on both 32 and 64-bit
+	 * systems, so we cant just use a pointer.
+	 */
+	tg_shadow_fb = (uintptr_t)gfx_state->tg_shadow_fb;
+	if (tg_shadow_fb)
+		BS->FreePages(tg_shadow_fb,
 		    gfx_state->tg_shadow_sz);
 	gfx_state->tg_shadow_sz =
 	    EFI_SIZE_TO_PAGES(efifb.fb_height * efifb.fb_width *
 	    sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));
 	status = BS->AllocatePages(AllocateMaxAddress, EfiLoaderData,
 	    gfx_state->tg_shadow_sz,
-	    (EFI_PHYSICAL_ADDRESS *)&gfx_state->tg_shadow_fb);
-	if (status != EFI_SUCCESS)
-		gfx_state->tg_shadow_fb = NULL;
+	    &tg_shadow_fb);
+	gfx_state->tg_shadow_fb = status == EFI_SUCCESS ?
+	    (uint32_t *)(uintptr_t)tg_shadow_fb : NULL;
```

Comment:
If this is above 4G, I think this truncates in the using the 32-bit loader, no?

Reply:
If `AllocatePages` was called with `AllocateAnyPages`, then that would be true yes. However, to my surprise, its called with `AllocateMaxAddress` which means that its treats the contents of `tg_shadow_fb` as the highest address it could give us. Since `gfx_state->tg_shadow_fb` starts out as NULL, the `AllocatePages` call fails every time...
This looks like it has been broken since 221376db0cb77035e325baf8b54d6224925b41f8

The following fixes it and insures that `ptr` is below 4G:
```
EFI_PHYSICAL_ADDRESS ptr = UINT32_MAX;

...

if (gfx_state->tg_shadow_fb != NULL)
	BS->FreePages((EFI_PHYSICAL_ADDRESS)gfx_state->tg_shadow_fb,
	    gfx_state->tg_shadow_sz);
gfx_state->tg_shadow_sz =
    EFI_SIZE_TO_PAGES(efifb.fb_height * efifb.fb_width *
    sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));

/*
 * EFI_PHYSICAL_ADDRESS is 8 bytes wide on both 32 and 64-bit
 * systems, so we cant just use a pointer.
 */
status = BS->AllocatePages(AllocateMaxAddress, EfiLoaderData,
    gfx_state->tg_shadow_sz, &ptr);
gfx_state->tg_shadow_fb = status == EFI_SUCCESS ?
    (uint32_t *)(uintptr_t)ptr : NULL;

return (0);
```

It looks like what you're describing also happens with `heap` in `efi_main.c`.

Nice catch

Reply:
Hmm, doing some reading, the UEFI spec in section 2.3.2.1 says:
>All descriptors are set to their 4GiB limits so that all of memory is accessible from all segments.

I guess that implies that, within the memory controlled by UEFI at least, `AllocatePages` can't actually give us an address over 2^30?

Though it also says in section 2.3.2:
>Paging mode may be enabled. If paging mode is enabled, PAE (Physical Address Extensions)
mode is recommended.

So IMO it's a safe bet to do this anyways to avoid overflows on any weird implementations.